peps/pep-0705.rst

```diff
@@ -185,111 +381,180 @@ Discussion:
+
+If all keys on a TypedDict should be read-only, the ``readonly`` flag can be used as a shorthand. *insert example*
 
 Reference Implementation
 ========================
 
-No reference implementation exists yet.
-
+No complete reference implementation exists yet. pyright 1.1.310 ships with a partial implementation of the ReadOnly qualifier.
 
 Rejected Alternatives
 =====================
 
-Several variations were considered and discarded:
+A TypedMapping protocol type
+----------------------------
+
+:pep:`705` proposed a ``TypedMapping`` protocol type, behaving much like a read-only TypedDict but without the constraint that the runtime type be a ``dict``. The behavior described in this PEP could then be obtained by inheriting a TypedDict from a TypedMapping. This has been set aside for now as more complex, without a strong use-case motivating the additional complexity.
+
+A higher-order Readonly type
+----------------------------
 
-* A ``readonly`` parameter to ``TypedDict``, behaving much like ``TypedMapping`` but with the additional constraint that instances must be dictionaries at runtime. This was discarded as less flexible due to the extra constraint; additionally, the new type nicely mirrors the existing ``Mapping``/``Dict`` types.
-* Inheriting from a ``TypedMapping`` subclass and ``TypedDict`` resulting in mutator methods being added for all fields, not just those actively (re)declared in the class body. Discarded as less flexible, and not matching how inheritance works in other cases for ``TypedDict`` (e.g. total=False and total=True do not affect fields not specified in the class body).
-* A generic type that removes mutator methods from its parameter, e.g. ``Readonly[MovieRecord]``. This would naturally want to be defined for a wider set of types than just ``TypedDict`` subclasses, and also raises questions about whether and how it applies to nested types. We decided to keep the scope of this PEP narrower.
-* Declaring methods directly on a ``TypedMapping`` class. Methods are a kind of property, but declarations on a ``TypedMapping`` class are defining keys, so mixing the two is potentially confusing. Banning methods also makes it very easy to decide whether a ``TypedDict`` subclass can mix in a protocol or not (yes if it's just ``TypedMapping`` superclasses, no if there's a ``Protocol``).
+A generalized higher-order type could be added that removes mutator methods from its parameter, e.g. ``ReadOnly[MovieRecord]``. For a TypedDict, this would be like adding ``readonly=True`` to the declaration. This would naturally want to be defined for a wider set of types than just TypedDict subclasses, and also raises questions about whether and how it applies to nested types. We decided to keep the scope of this PEP narrower.
 
+Preventing other keys with the typing.final decorator
```

Comment:
When I saw the `other_keys` part of the PEP I thought of the proposal in https://discuss.python.org/t/pep-692-using-typeddict-for-more-precise-kwargs-typing/17314/87?u=jelle, which has come up a few times. It's a way to specify that the remaining keys in a TypedDict are of a particular type. I have two concerns:
- This PEP's `other_keys` flag could lead to user confusion if users expect it to work like the linked idea
- This PEP's use of the `other_keys` name could make it more difficult for a future PEP to implement the other idea, if that PEP also wants to use the name "other_keys".

Reply:
I think the other proposal could reuse this argument:

`other_keys=int`

They're both constraining the "other keys", and they're mutually incompatible, so I think this would be ok?

Reply:
#3441 was just created, proposing an `__extra__: T` field. I feel if both PEPs are accepted as currently written, we'd end up in a confusing situation, but I'll have to think more about how I'd recommend combining the two.

Reply:
I'm happy for this PEP to defer the exact way of specifying the extra field to PEP-728 (and associated discussion). My only concern is if it lands first!

* They could both use `other_keys`, and `False` could be a synonym for `Never`
* They could both use `other_keys`, and this PEP could add *only* `other_keys=Never` (i.e. it's no longer a boolean, but if only this PEP merges, no other type is allowed)

(Replace `other_keys` with `__extra__` or whatever comes out of the PEP-728 discussion.)

Reply:
Is this something I should note directly in the PEP draft?

Reply:
`other_keys=Never` sounds like a reasonable approach.